4 analyzer

1EVILGUN1 opened the pull request.

1EVILGUN1 committed:
feat: create module Analyzer and rebuilding the project

Создание модуля Analyzer для анализа поведения датчиков и подготовка сценария записанного в базу данных. Также вынос моделей в отдельный модуль event-model для оптимального чтения модуля collector. Добавление еще одной схемы контроллера для хабов hub_router_controller

1EVILGUN1 committed:
fix reformat code

чистка кода

1EVILGUN1 committed:
fix rebuilding pom

1EVILGUN1 committed:
upd pom aggregator

1EVILGUN1 committed:
fix rebuilding pom telemetry

пересборка модулей temetry

EugeneLenkevich commented on telemetry/aggregator/src/main/java/ru/yandex/practicum/serialize/SensorsSnapshotSrializer.java:
```diff
@@ -0,0 +1,48 @@
+
+            byte[] bytes = outputStream.toByteArray();
+
+            log.info("Данные сериализованы в формат Avro:\n{}", hexFormat.formatHex(bytes));
```
А что тут будет выведено, поможет ли нам потом это при отладке, может тут какие то идентификаторы события лучше вывести? :)

EugeneLenkevich commented on telemetry/analyzer/src/main/java/ru/yandex/practicum/HubEventProcessor.java:
```diff
@@ -0,0 +1,135 @@
+                new OffsetAndMetadata(record.offset() + 1)
+        );
+
+        if (count % 10 == 0) {
```
count % 10 == 0

Если честно так не нравятся такие условия в коде, это как то не по человечески (в смысле не по кожаному), мы все таки лучше воспринимает слова, а не наборы символов. А еще это детали, которые могут измениться
Давай спрячем их в метод, например isSynsConditionsReached - у нас появится нормальная семантика, код как текст читать можно будет :)

EugeneLenkevich commented on telemetry/analyzer/src/main/java/ru/yandex/practicum/HubEventProcessor.java:
```diff
@@ -0,0 +1,135 @@
+
+
+                        switch (record.value().getPayload().getClass().getSimpleName()) {
+                            case "ScenarioRemovedEventAvro":
```
"ScenarioRemovedEventAvro"

Допустим, когда то потом мы будем делать рефакторинг, переименуем класс, но эту строку конечно же пропустим и потом будем долго гадать почему у нас все сломалось :)
В общем от строк лучше избавится каким то образом.

Можно запилить паттерн стратегия. Если в кратце то создается мапа класс=>обработчик, далее, приходит сообщение, по его классу мы вытаскиваем нужный обработчик. В обработчике содержится логика из этих case-блоков

EugeneLenkevich commented on telemetry/analyzer/src/main/java/ru/yandex/practicum/SnapshotProcessor.java:
```diff
@@ -0,0 +1,168 @@
+        if (count % 10 == 0) {
+            consumer.commitAsync(currentOffsets, (offsets, exception) -> {
+                if (exception != null) {
+                    log.warn("Ошибка во время фиксации оффсетов: {}", offsets, exception);
```
Только log.error наверно  уж :)

EugeneLenkevich commented on telemetry/analyzer/src/main/java/ru/yandex/practicum/SnapshotProcessor.java:
```diff
@@ -0,0 +1,168 @@
+                new OffsetAndMetadata(record.offset() + 1)
+        );
+
+        if (count % 10 == 0) {
```
count % 10 == 0

Вот, кстати, это же у нас уже было. То есть тут кандидат на вынос в отельный метод и может даже отдельного бина. А еще вдруг это значение, 10, мы захотим вынести в настройки. А давай так и сделаем - сделаем бин, проверяющий условие, а 10 вынесем его в настройки, те что в application.yml

EugeneLenkevich commented on telemetry/analyzer/src/main/java/ru/yandex/practicum/SnapshotProcessor.java:
```diff
@@ -0,0 +1,168 @@
+                        state.setTimestamp(stateAvro.getTimestamp());
+
+                        switch (stateAvro.getData().getClass().getSimpleName()) {
+                            case "MotionSensorAvro":
```
Тут тоже ненадежно, в идеале, конечно, тестами надо бы покрыт, но если время будет

EugeneLenkevich commented on telemetry/analyzer/src/main/java/ru/yandex/practicum/service/HubService.java:
```diff
@@ -0,0 +1,96 @@
+        if (!sensorRepository.existsById(event.getId())) {
+            SensorEntity sensorEntity = new SensorEntity(event.getId(), event.getHubId());
+            sensorRepository.save(sensorEntity);
+            System.out.println("Добавлено устройство");
```
System.out.println

Лучше выводить все через логгер, log.info(...)
Этот вывод может потеряться в зависимости от настроек логирования

EugeneLenkevich commented on telemetry/analyzer/src/main/java/ru/yandex/practicum/service/HubService.java:
```diff
@@ -0,0 +1,96 @@
+        }
+    }
+
+    public void processingEvent(DeviceRemovedEvent event) {
```
Тут есть удаление, наверное тое стоит повесить `@Transactional`?
И проверь, пожалуйста, другие публичные методы сервисов, везде ли есть транзакция где меняются данные (insert, update, delete)

EugeneLenkevich commented on telemetry/analyzer/src/main/java/ru/yandex/practicum/service/HubService.java:
```diff
@@ -0,0 +1,96 @@
+
+        for (ScenarioCondition sc : event.getConditions()) {
+            ConditionEntity conditionEntity = new ConditionEntity();
+            conditionEntity.setType(sc.getType().name());
```
conditionEntity.setType(sc.getType().name());

А в сущности же можно сделать поле типа enum, будет проще и надежней чем строки хранить :)

EugeneLenkevich commented on telemetry/analyzer/src/main/java/ru/yandex/practicum/service/HubService.java:
```diff
@@ -0,0 +1,96 @@
+            actionRepository.save(actionEntity);
+            Optional<SensorEntity> sensorEntity = sensorRepository.findByIdAndHubId(da.getSensorId(), event.getHubId());
+            if (sensorEntity.isEmpty()) {
+                System.out.println("Нужный сенсор не найден");
```
log.warn

EugeneLenkevich commented on telemetry/collector/src/main/java/ru/yandex/practicum/controller/CollectorController.java:
```diff
@@ -7,35 +7,33 @@
-        switch (event.getType()) {
-            case SensorEventType.CLIMATE_SENSOR_EVENT:
+        switch (event.getClass().getSimpleName()) {
+            case "ClimateSensorEvent":
```
Снова строки :)

EugeneLenkevich commented on telemetry/collector/src/main/java/ru/yandex/practicum/grpc/EventController.java:
```diff
@@ -19,13 +21,19 @@
+        this.hubEventHandlers = hubEventHandlers.stream()
+                .collect(Collectors.toMap(
+                        HubEventHandler::getMessageType,
+                        Function.identity()
```
Вот, кстати, стратегия у нас тут уже есть, как пример реализации :)

EugeneLenkevich commented on telemetry/analyzer/src/main/java/ru/yandex/practicum/model/ConditionEntity.java:
```diff
@@ -0,0 +1,20 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    private String type;
+    private String operation;
```
private String type;
    private String operation;

А что то из этого нельзя сделать enum?
Строки все таки ненадежно. Хотя в базе можно и строку хранить, смотря как настроить JPA

https://www.baeldung.com/jpa-persisting-enums-in-jpa

![image](https://github.com/user-attachments/assets/5fd661ff-2a75-47b6-9129-f036310f0a1e)

EugeneLenkevich reviewed (changes requested):
Привет, Яков! Спасибо за PR.

Предложу ряд правок, если есть время, но главное, два момента
- заменить System.out.println на log.info/log.warn/log.error
- избавится от строк в перечислениях любым способом, такой код со строками  рефакторить будет просто кромешный ад, особенно с учетом что тестов нет. Одно переименование и у нас куча всего отвалится

Но, в целом, отличная работа, надо только причесать немного :)

1EVILGUN1 committed:
fix comments

исправлены замечания

EugeneLenkevich commented on telemetry/analyzer/src/main/java/ru/yandex/practicum/SnapshotProcessor.java:
```diff
@@ -69,58 +68,47 @@ public void start() {
-
-                                state.setData(switchSensorEvent);
-                                break;
+                        if (stateAvro.getData() instanceof MotionSensorAvro) {
```
Кстати, есть еще такой вариант, с 17ой версии
https://docs.oracle.com/en/java/javase/21/language/pattern-matching-switch.html#GUID-E69EEA63-E204-41B4-AA7F-D58B26A3B232

EugeneLenkevich reviewed (approved):
Ну красота же, теперь все отлично, спасибо :)

1EVILGUN1 committed:
fix add in application setting consumer and producer kafka

вынос в application продюсер и консьмер для оптимизации и удобной настройки из отдельного файла. Также сделаны обработки ошибок классов handle
